driver/sum_driver.hpp

```diff
@@ -0,0 +1,334 @@
+
+    in      = std::vector<Tgpu>(in_sz, static_cast<Tgpu>(0));
+    out     = std::vector<Tgpu>(out_sz, static_cast<Tgpu>(0));
+    outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+
+    int status;
+
+    for(int i = 0; i < in_sz; i++)
+    {
+        in[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(0.0), static_cast<Tgpu>(1.0));
+    }
+    status = in_dev->ToGPU(q, in.data());
+
+    status |= out_dev->ToGPU(q, out.data());
+
+    if(status != 0)
+        std::cout << "Error copying data to GPU\n" << std::endl;
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int SumDriver<Tgpu, Tref>::RunForwardGPU()
+{
+    float kernel_total_time = 0.0;
+    float kernel_first_time = 0.0;
+
+    Timer t;
+    START_TIME
+
+    for(int i = 0; i < inflags.GetValueInt("iter"); i++)
+    {
+        miopenSumForward(GetHandle(),
+                         nanPropagation,
+                         workspace_dev->GetMem(),
+                         ws_sizeInBytes,
+                         inputDesc,
+                         in_dev->GetMem(),
+                         dim,
+                         outputDesc,
+                         out_dev->GetMem());
+
+        float time = 0.0;
+        miopenGetKernelTime(GetHandle(), &time);
+        kernel_total_time += time;
+        if(i == 0)
+            kernel_first_time = time;
+    }
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        STOP_TIME
+        int iter = inflags.GetValueInt("iter");
+        if(WALL_CLOCK)
+            printf("Wall-clock Time Forward Sum Elapsed: %f ms\n", t.gettime_ms() / iter);
+
+        float kernel_average_time =
+            iter > 1 ? (kernel_total_time - kernel_first_time) / (iter - 1) : kernel_first_time;
+        printf("GPU Kernel Time Forward Sum Elapsed: %f ms\n", kernel_average_time);
+    }
+
+    out_dev->FromGPU(GetStream(), out.data());
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int SumDriver<Tgpu, Tref>::RunForwardCPU()
+{
+    mloSumForwardRunHost<Tgpu, Tref>(
+        inputDesc, outputDesc, in.data(), outhost.data(), dim, nanPropagation);
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int SumDriver<Tgpu, Tref>::RunBackwardGPU()
+{
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+Tref SumDriver<Tgpu, Tref>::GetTolerance()
+{
+    if(data_type == miopenHalf)
+    {
+        return 1e-3;
+    }
+    else if(data_type == miopenFloat)
+    {
+        return 5e-5;
+    }
+    else if(data_type == miopenBFloat16)
+    {
+        return 1e-2;
+    }
+    return 0;
+}
+
+template <typename Tgpu, typename Tref>
+int SumDriver<Tgpu, Tref>::VerifyForward()
+{
+    RunForwardCPU();
+    const Tref tolerance = GetTolerance();
+    auto error           = miopen::rms_range(outhost, out);
+
+    if(!std::isfinite(error) || error > tolerance)
+    {
+        std::cout << "Forward Sum FAILED: " << error << std::endl;
```

Comment:
Print tolerance (everywhere)
```suggestion
        std::cout << "Forward Sum FAILED: " << error << " > " << tolerance << std::endl;
```
[Notice] `cout` is fine here because this is not a driver's failure (driver is working normally, it's a problem in the library).

Reply:
fixed it

Reply:
[resolved]